object_intern.go

```diff
@@ -100,49 +103,88 @@ func (oi *ObjectIntern) DecompressString(in string) (string, error) {
 // If the object is added to the store its reference count is set to 1.
 func (oi *ObjectIntern) AddOrGet(obj []byte, safe bool) (uintptr, error) {
-	var addr gos.ObjAddr
-	var ok bool
-	var err error
-	var objString string
 
-	objComp := obj
+	// if either of these two terms is true then the rest of this block
+	// requires a lot of allocations
+	if (oi.conf.Compression != None) || (safe && oi.conf.Compression == None) {
+		objComp := obj
 
-	if oi.conf.Compression != None {
-		objComp = oi.compress(obj)
-	}
+		if oi.conf.Compression != None {
+			// this returns a new byte slice, so we don't need to check for safe
+			objComp = oi.compress(obj)
+		}
 
-	// the only case we need to handle specially is when compression
-	// is turned off and the user has requested that the operation be safe
-	if safe && oi.conf.Compression == None {
-		// create a copy so we don't modify the original []byte
-		objComp = make([]byte, len(obj))
-		copy(objComp, obj)
-	}
+		// the only case we need to handle specially is when compression
+		// is turned off and the user has requested that the operation be safe
+		if safe && oi.conf.Compression == None {
+			// create a copy so we don't modify the original []byte
+			objComp = make([]byte, len(obj))
+			copy(objComp, obj)
+		}
 
-	// string used for the index
-	objString = string(objComp)
+		// acquire lock
+		oi.Lock()
+
+		// try to find the object in the index
+		addr, ok := oi.objIndex[string(objComp)]
+		if ok {
+			// increment reference count by 1
+			(*(*uint32)(unsafe.Pointer(addr + uintptr(len(objComp)))))++
+			oi.Unlock()
+			return addr, nil
+		}
+
+		// The object is not in the index therefore it is not in the store.
+		// We need to set its initial reference count to 1 before adding it.
+		//
+		// The object store backend has no knowledge of a reference count, so
+		// we need to manage it at this layer. Here we add 4 bytes to be used
+		// henceforth as the reference count for this object. Reference count is
+		// always placed as the LAST 4 bytes of an object and is NEVER compressed.
+		objComp = append(objComp, InitialRefCount...)
+		addr, err := oi.store.Add(objComp)
+		if err != nil {
+			oi.Unlock()
+			return 0, err
+		}
+
+		// remove ref count from end before converting to string
+		objString := string(objComp[:len(objComp)-4])
+
+		// set objString data to the object inside the object store
+		((*reflect.StringHeader)(unsafe.Pointer(&objString))).Data = addr
+
+		// add the object to the index
+		oi.objIndex[objString] = addr
+
+		oi.Unlock()
+		return addr, nil
+	}
 
+	// if neither of those terms is true then we can avoid costly allocations
 	// acquire lock
 	oi.Lock()
 
 	// try to find the object in the index
-	addr, ok = oi.objIndex[objString]
+	addr, ok := oi.objIndex[string(obj)]
 	if ok {
 		// increment reference count by 1
-		(*(*uint32)(unsafe.Pointer(addr + uintptr(len(objComp)))))++
+		(*(*uint32)(unsafe.Pointer(addr + uintptr(len(obj)))))++
```

Comment:
since it takes me a min to understand this line every time i look at it, could we put that into a separate function like `AddToRefCount(addr, len uintptr, delta uint32)` or something like that? `delta` of `-1` can be specified as `^uint32(c-1)` like in `atomic`: https://golang.org/pkg/sync/atomic/#AddUint32
Side note: actually this operation can be atomic instead of acquiring the write lock

Reply:
Yeah that looks cool, I'll check it out. I just wonder if those atomic operations will work with MMapped memory?

Reply:
I'd assume that anything which is a pointer to a uint32 would work

Reply:
They do appear to work, but from my initial tests they cause more allocations. I will look into this further and get back to you when I have some solid answers / evidence.

Reply:
I can't put too much weight into the time changes, but that is significant enough that I think I won't use atomics for now.

```
benchmark                                       old ns/op      new ns/op      delta
BenchmarkAddOrGetCompressed10-12                2368           2815           +18.88%
BenchmarkAddOrGetCompressed100-12               21102          25280          +19.80%
BenchmarkAddOrGetCompressed1000-12              156915         177825         +13.33%
BenchmarkAddOrGetCompressed10000-12             1574681        1763613        +12.00%
BenchmarkAddOrGetCompressed100000-12            18482567       22232140       +20.29%
BenchmarkAddOrGetCompressed1000000-12           377701741      393682623      +4.23%
BenchmarkAddOrGetCompressed5000000-12           4116743009     4095563962     -0.51%
BenchmarkAddOrGetUnsafe10-12                    326            779            +138.96%
BenchmarkAddOrGetUnsafe100-12                   3375           4845           +43.56%
BenchmarkAddOrGetUnsafe1000-12                  37244          56699          +52.24%
BenchmarkAddOrGetUnsafe10000-12                 447499         694327         +55.16%
BenchmarkAddOrGetUnsafe100000-12                5398773        7848335        +45.37%
BenchmarkAddOrGetUnsafe1000000-12               156685284      206915016      +32.06%
BenchmarkAddOrGetUnsafe5000000-12               3306532892     3325429320     +0.57%
BenchmarkAddOrGetSafe10-12                      471            1014           +115.29%
BenchmarkAddOrGetSafe100-12                     5038           7216           +43.23%
BenchmarkAddOrGetSafe1000-12                    53682          77602          +44.56%
BenchmarkAddOrGetSafe10000-12                   654150         923191         +41.13%
BenchmarkAddOrGetSafe100000-12                  7685703        10947507       +42.44%
BenchmarkAddOrGetSafe1000000-12                 212833788      240168220      +12.84%
BenchmarkAddOrGetSafe5000000-12                 3600328614     3596351423     -0.11%
BenchmarkAddOrGetStringCompressed10-12          1607           2695           +67.70%
BenchmarkAddOrGetStringCompressed100-12         16966          20254          +19.38%
BenchmarkAddOrGetStringCompressed1000-12        169804         197809         +16.49%
BenchmarkAddOrGetStringCompressed10000-12       1836669        2010640        +9.47%
BenchmarkAddOrGetStringCompressed100000-12      20167352       22458518       +11.36%
BenchmarkAddOrGetStringCompressed1000000-12     385989993      412172366      +6.78%
BenchmarkAddOrGetStringCompressed5000000-12     4038263104     4078744896     +1.00%
BenchmarkAddOrGetUnsafeString10-12              338            794            +134.91%
BenchmarkAddOrGetUnsafeString100-12             3458           4778           +38.17%
BenchmarkAddOrGetUnsafeString1000-12            37051          58385          +57.58%
BenchmarkAddOrGetUnsafeString10000-12           469431         719560         +53.28%
BenchmarkAddOrGetUnsafeString100000-12          5514563        8266234        +49.90%
BenchmarkAddOrGetUnsafeString1000000-12         153315804      205854523      +34.27%
BenchmarkAddOrGetUnsafeString5000000-12         3286636130     3450282272     +4.98%
BenchmarkAddOrGetSafeString10-12                469            965            +105.76%
BenchmarkAddOrGetSafeString100-12               5158           6944           +34.63%
BenchmarkAddOrGetSafeString1000-12              54555          77939          +42.86%
BenchmarkAddOrGetSafeString10000-12             668555         949112         +41.96%
BenchmarkAddOrGetSafeString100000-12            7472386        10428552       +39.56%
BenchmarkAddOrGetSafeString1000000-12           214243376      234108750      +9.27%
BenchmarkAddOrGetSafeString5000000-12           3556085813     3500545906     -1.56%
BenchmarkCompressShoco-12                       340            338            -0.59%
BenchmarkDecompressShoco-12                     251            248            -1.20%
BenchmarkCompressNone-12                        2.76           2.77           +0.36%
BenchmarkDecompressNone-12                      2.81           2.80           -0.36%
BenchmarkCompressSzShoco-12                     195            195            +0.00%
BenchmarkDecompressSzShoco-12                   190            191            +0.53%
BenchmarkCompressSzNone-12                      2.59           2.59           +0.00%
BenchmarkDecompressSzNone-12                    2.69           2.66           -1.12%

benchmark                                       old allocs     new allocs     delta
BenchmarkAddOrGetCompressed10-12                10             10             +0.00%
BenchmarkAddOrGetCompressed100-12               100            100            +0.00%
BenchmarkAddOrGetCompressed1000-12              1000           1000           +0.00%
BenchmarkAddOrGetCompressed10000-12             10010          10010          +0.00%
BenchmarkAddOrGetCompressed100000-12            101070         101070         +0.00%
BenchmarkAddOrGetCompressed1000000-12           1356154        1356225        +0.01%
BenchmarkAddOrGetCompressed5000000-12           10304885       10304894       +0.00%
BenchmarkAddOrGetUnsafe10-12                    0              0              +0.00%
BenchmarkAddOrGetUnsafe100-12                   0              0              +0.00%
BenchmarkAddOrGetUnsafe1000-12                  0              0              +0.00%
BenchmarkAddOrGetUnsafe10000-12                 3              5              +66.67%
BenchmarkAddOrGetUnsafe100000-12                360            540            +50.00%
BenchmarkAddOrGetUnsafe1000000-12               106913         213864         +100.04%
BenchmarkAddOrGetUnsafe5000000-12               5306538        5305752        -0.01%
BenchmarkAddOrGetSafe10-12                      10             10             +0.00%
BenchmarkAddOrGetSafe100-12                     100            100            +0.00%
BenchmarkAddOrGetSafe1000-12                    1000           1000           +0.00%
BenchmarkAddOrGetSafe10000-12                   10010          10010          +0.00%
BenchmarkAddOrGetSafe100000-12                  101035         101035         +0.00%
BenchmarkAddOrGetSafe1000000-12                 1413620        1413796        +0.01%
BenchmarkAddOrGetSafe5000000-12                 15303922       15303623       -0.00%
BenchmarkAddOrGetStringCompressed10-12          20             20             +0.00%
BenchmarkAddOrGetStringCompressed100-12         200            200            +0.00%
BenchmarkAddOrGetStringCompressed1000-12        2000           2000           +0.00%
BenchmarkAddOrGetStringCompressed10000-12       20010          20010          +0.00%
BenchmarkAddOrGetStringCompressed100000-12      202140         202141         +0.00%
BenchmarkAddOrGetStringCompressed1000000-12     2356164        2356145        -0.00%
BenchmarkAddOrGetStringCompressed5000000-12     15307328       15305693       -0.01%
BenchmarkAddOrGetUnsafeString10-12              0              0              +0.00%
BenchmarkAddOrGetUnsafeString100-12             0              0              +0.00%
BenchmarkAddOrGetUnsafeString1000-12            0              0              +0.00%
BenchmarkAddOrGetUnsafeString10000-12           3              5              +66.67%
BenchmarkAddOrGetUnsafeString100000-12          360            540            +50.00%
BenchmarkAddOrGetUnsafeString1000000-12         107050         214096         +100.00%
BenchmarkAddOrGetUnsafeString5000000-12         5305711        5307816        +0.04%
BenchmarkAddOrGetSafeString10-12                10             10             +0.00%
BenchmarkAddOrGetSafeString100-12               100            100            +0.00%
BenchmarkAddOrGetSafeString1000-12              1000           1000           +0.00%
BenchmarkAddOrGetSafeString10000-12             10010          10010          +0.00%
BenchmarkAddOrGetSafeString100000-12            101035         101035         +0.00%
BenchmarkAddOrGetSafeString1000000-12           1413706        1413736        +0.00%
BenchmarkAddOrGetSafeString5000000-12           15303671       15303504       -0.00%
BenchmarkCompressShoco-12                       1              1              +0.00%
BenchmarkDecompressShoco-12                     1              1              +0.00%
BenchmarkCompressNone-12                        0              0              +0.00%
BenchmarkDecompressNone-12                      0              0              +0.00%
BenchmarkCompressSzShoco-12                     3              3              +0.00%
BenchmarkDecompressSzShoco-12                   3              3              +0.00%
BenchmarkCompressSzNone-12                      0              0              +0.00%
BenchmarkDecompressSzNone-12                    0              0              +0.00%

benchmark                                       old bytes      new bytes      delta
BenchmarkAddOrGetCompressed10-12                1120           1120           +0.00%
BenchmarkAddOrGetCompressed100-12               11200          11200          +0.00%
BenchmarkAddOrGetCompressed1000-12              112013         112013         +0.00%
BenchmarkAddOrGetCompressed10000-12             1121068        1121067        -0.00%
BenchmarkAddOrGetCompressed100000-12            11292862       11292813       -0.00%
BenchmarkAddOrGetCompressed1000000-12           159547258      159561872      +0.01%
BenchmarkAddOrGetCompressed5000000-12           1146566088     1146570696     +0.00%
BenchmarkAddOrGetUnsafe10-12                    0              0              +0.00%
BenchmarkAddOrGetUnsafe100-12                   0              0              +0.00%
BenchmarkAddOrGetUnsafe1000-12                  2              4              +100.00%
BenchmarkAddOrGetUnsafe10000-12                 382            577            +51.05%
BenchmarkAddOrGetUnsafe100000-12                32670          49045          +50.12%
BenchmarkAddOrGetUnsafe1000000-12               14097934       28203440       +100.05%
BenchmarkAddOrGetUnsafe5000000-12               582816808      582585784      -0.04%
BenchmarkAddOrGetSafe10-12                      80             80             +0.00%
BenchmarkAddOrGetSafe100-12                     800            800            +0.00%
BenchmarkAddOrGetSafe1000-12                    8004           8007           +0.04%
BenchmarkAddOrGetSafe10000-12                   152720         152719         -0.00%
BenchmarkAddOrGetSafe100000-12                  1656864        1656871        +0.00%
BenchmarkAddOrGetSafe1000000-12                 50580016       54410028       +7.57%
BenchmarkAddOrGetSafe5000000-12                 822280360      822218168      -0.01%
BenchmarkAddOrGetStringCompressed10-12          1200           1200           +0.00%
BenchmarkAddOrGetStringCompressed100-12         12000          12000          +0.00%
BenchmarkAddOrGetStringCompressed1000-12        120013         120013         +0.00%
BenchmarkAddOrGetStringCompressed10000-12       1272991        1272987        -0.00%
BenchmarkAddOrGetStringCompressed100000-12      12969763       12970029       +0.00%
BenchmarkAddOrGetStringCompressed1000000-12     174716538      174712725      -0.00%
BenchmarkAddOrGetStringCompressed5000000-12     1222380232     1221986888     -0.03%
BenchmarkAddOrGetUnsafeString10-12              0              0              +0.00%
BenchmarkAddOrGetUnsafeString100-12             0              0              +0.00%
BenchmarkAddOrGetUnsafeString1000-12            2              4              +100.00%
BenchmarkAddOrGetUnsafeString10000-12           384            580            +51.04%
BenchmarkAddOrGetUnsafeString100000-12          32675          48980          +49.90%
BenchmarkAddOrGetUnsafeString1000000-12         14135241       28267808       +99.98%
BenchmarkAddOrGetUnsafeString5000000-12         620460712      621072936      +0.10%
BenchmarkAddOrGetSafeString10-12                80             80             +0.00%
BenchmarkAddOrGetSafeString100-12               800            800            +0.00%
BenchmarkAddOrGetSafeString1000-12              8004           8007           +0.04%
BenchmarkAddOrGetSafeString10000-12             152721         152721         +0.00%
BenchmarkAddOrGetSafeString100000-12            1656718        1656781        +0.00%
BenchmarkAddOrGetSafeString1000000-12           50597776       50604182       +0.01%
BenchmarkAddOrGetSafeString5000000-12           822228296      822193560      -0.00%
BenchmarkCompressShoco-12                       112            112            +0.00%
BenchmarkDecompressShoco-12                     112            112            +0.00%
BenchmarkCompressNone-12                        0              0              +0.00%
BenchmarkDecompressNone-12                      0              0              +0.00%
BenchmarkCompressSzShoco-12                     136            136            +0.00%
BenchmarkDecompressSzShoco-12                   136            136            +0.00%
BenchmarkCompressSzNone-12                      0              0              +0.00%
BenchmarkDecompressSzNone-12                    0              0              +0.00%
```